### ISSUE TITLE
Godot 4 Visuals Shaders are sensitive to System Scale in size, but their position remains the same and distorted

### ISSUE THREAD

**Issue by ArseniyMirniy** [godotengine/godot#98513]
### Tested versions

Godot 4.3 Stable

### System information

Windows 10 64, Godot 4.3 Stable, Forward+

### Issue description

Godot Engine does not respect the grid of visual nodes (in shaders or other graphs) if the UI scale has changed. It is crucial for public projects or games developed by multiple people on different screens. Or even if a single person tries to work on a desktop (100% scale) and laptop (150-200% scale).

### You made a node graph, when your system is at 100% scale, it looks clean:
![Godot_Visual_Shaders_Bug](https://github.com/user-attachments/assets/dfd5555c-0217-48e9-aa3e-b8b0ee076238)

### With 150% scale, nodes are mashed together:
![Godot_Visual_Shaders_Bug2](https://github.com/user-attachments/assets/e068ead6-c6cf-42b6-9f17-f985f5b920ba)

### With 200% scale, it gets even worse:
![Godot_Visual_Shaders_Bug3](https://github.com/user-attachments/assets/78da3312-59b0-4cc4-835c-93de1459e808)

The solution is not just change sizes of nodes themselves, but also make them respect the grid (and reorganize accordingly).

### Steps to reproduce

1. Use system scale of 100% (and/or editor scale of 100%)
2. Make a visual shader with several nodes aligned in a grid, save it
3. Change UI scale (in the system or in the Godot Editor), relaunch
4. Grid is now messed up completely

### Minimal reproduction project (MRP)

You can try my graph from here: https://github.com/ArseniyMirniy/Godot-4-Free-Color-Correction-and-Screen-Effects-Visual-Shader

**Comment by Calinou** [comment]
This should likely be fixed by never scaling anything drawn within GraphNode/GraphEdit in the editor theme, and having only the default zoom level change instead (i.e. what appears as 100% zoom). The same approach is used for the 2D editor viewport already.

**Comment by ArseniyMirniy** [comment]
> This should likely be fixed by never scaling anything drawn within GraphNode/GraphEdit in the editor theme, and having only the default zoom level change instead (i.e. what appears as 100% zoom). The same approach is used for the 2D editor viewport already.
> 
> GraphNode/GraphEdit use fully vector-based drawing (including MSDF fonts), so they will still look good at high zoom levels.

1. If so, is the fix a relatively fast with tweaking a few settings in one place and adding the scale factor to zoom level in another? If so, can it be expected in 4.4 or 4.5 at least?

2. Off-topic: thanks for AgX, it's a cool addition to the engine, can't wait to see 4.4 Stable

**Comment by Calinou** [comment]
> 1. If so, is the fix a relatively fast with tweaking a few settings in one place and adding the scale factor to zoom level in another? If so, can it be expected in 4.4 or 4.5 at least? I will be glad to test!

Yes and no. There's still the difficult problem of handling editor icons, as the visual shader editor will draw icons within its GraphEdit (e.g. for checkboxes or type icons). These are rendered at the editor scale, so if we discard the scaling on everything but the icons, the icons would look much larger than the rest on hiDPI displays.

**Comment by KoB-Kirito** [comment]
"clean"
Lines that actual don't overlap or are moveable would be clean. New GraphEdit when?

**Comment by Calinou** [comment]
> "clean"
Lines that actual don't overlap or are moveable would be clean. New GraphEdit when?

This is off-topic for this bug report. Feature proposals should be made on https://github.com/godotengine/godot-proposals.

**Comment by YeldhamDev** [comment]
From what I tested, it seems that this problem has been mostly mitigated by #97620. On extreme scales, some offsets are a little wonky doing to diverging minimum sizes of elements inside nodes, but the graph itself doesn't seem to be at fault here.

The only remaining big issue is how the grid snapping is treated differently depending on the scale, which I will open a PR for.
